File: UnityApplication/app/src/main/java/com/example/testUnity/BackgroundCheckWorker.java
package com.example.testUnity;

import android.content.Context;
import android.os.BatteryManager;
import android.os.Build;
import android.util.Log;

import androidx.work.Worker;
import androidx.work.WorkerParameters;
import com.microsoft.signalr.HubConnection;
import com.microsoft.signalr.HubConnectionBuilder;
import com.microsoft.signalr.OnClosedCallback;
import java.util.*;
import io.reactivex.rxjava3.functions.Action;

public class BackgroundCheckWorker extends Worker
{
    private static final String TAG = "BACKGROUND CHECK";

    private String url = "http://192.168.50.61:5109/Connection";
    private HubConnection hubConnection = HubConnectionBuilder.create(url).build();

    public BackgroundCheckWorker(Context context, WorkerParameters workerParams) {
        super(context, workerParams);
    }

    public Result doWork() {
        //Check battery
        int batteryLevel = getPicoBatteryLevel();
        BatteryManager batteryManager = (BatteryManager) getApplicationContext().getSystemService(Context.BATTERY_SERVICE);
        int batteryStatus = 0;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            batteryStatus = batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_STATUS);
        }

        // Send battery level to API
        sendBatteryLevelToApiUsingSignalR(batteryLevel, batteryStatus);
        return Result.success();
    }

    private int getPicoBatteryLevel() {
        // Retrieve the headset's battery level
        return getBatteryPercentage(getApplicationContext());
    }

    private int getBatteryPercentage(Context context) {
        BatteryManager batteryManager = (BatteryManager) getApplicationContext().getSystemService(Context.BATTERY_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
        }
        return 0;
    }

    private void sendBatteryLevelToApiUsingSignalR(int batteryLevel, int batteryStatus)
    {
        Log.d(TAG, "BACKGROUND SERVICE CHECKER.");

        hubConnection.onClosed(new OnClosedCallback() {
            @Override
            public void invoke(Exception exception) {
                try {
                    Thread.sleep(new Random().nextInt(5) * 1000L);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
                hubConnection.start();
            }
        });

        hubConnection.start().doOnComplete(new Action() {
            @Override
            public void run() throws Throwable {
                try {
                    hubConnection.invoke("SendBatteryStats", batteryLevel);
                } catch(Exception e)
                {
                    System.out.print("Error sending stats: $ex");
                }
                finally {
                    hubConnection.stop();
                }
            }
        }).blockingAwait();
    }
}

File: TestKotlin/app/src/main/java/com/example/testkotlin/BackgroundCheckWorker.kt
package com.example.testkotlin

import android.content.Context
import android.os.BatteryManager
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.microsoft.signalr.HubConnection
import com.microsoft.signalr.HubConnectionBuilder
import java.util.*

class BackgroundCheckWorker(context: Context, workerParams: WorkerParameters) :
    Worker(context, workerParams)
{
    private val url = "http://196.168.210.233:5109/Connection"
    private val hubConnection: HubConnection = HubConnectionBuilder.create(url).build()

    override fun doWork(): Result {

        //Check battery
        val batteryLevel = getPicoBatteryLevel()
        val batteryManager = applicationContext?.getSystemService(Context.BATTERY_SERVICE) as BatteryManager
        val batteryStatus = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            batteryManager.getIntProperty(BatteryManager.BATTERY_PROPERTY_STATUS)
        } else {
            TODO("VERSION.SDK_INT < O")
        }

        // Send battery level to API
        sendBatteryLevelToApiUsingSignalR(batteryLevel, batteryStatus)
        return Result.success()
    }

    private fun getPicoBatteryLevel(): Int? {
        // Retrieve the headset's battery level
        return getBatteryPercentage(applicationContext)
    }

    private fun getBatteryPercentage(context: Context?): Int? {
        val bm = context?.getSystemService(AppCompatActivity.BATTERY_SERVICE) as BatteryManager
        return bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
    }

    private fun sendBatteryLevelToApiUsingSignalR(batteryLevel: Int?, batteryStatus: Int) {
        hubConnection.onClosed {
            println("Connection closed: ${it.message}")
            // Attempt to restart the connection
            Thread.sleep(Random().nextInt(5) * 1000L)
            hubConnection.start()
        }

       val e = hubConnection.start().doOnComplete {
            try {
                hubConnection.invoke("SendBatteryStats", batteryLevel)
                println("Stats sent")
            } catch (ex: Exception) {
                println("Error sending stats: $ex")
            } finally {
                hubConnection.stop()
                println("Connection stopped")
            }
        }.blockingAwait()
    }
}